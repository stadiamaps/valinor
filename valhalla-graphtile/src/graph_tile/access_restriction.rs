use crate::graph_tile::GraphTileError;
use crate::Access;
use bitfield_struct::bitfield;
use enumset::EnumSet;
use zerocopy::try_transmute;
use zerocopy_derive::{FromBytes, TryFromBytes};

/// Types of access restrictions.
#[derive(TryFromBytes, Debug, Eq, PartialEq)]
#[repr(u8)]
pub enum AccessRestrictionType {
    Hazmat,
    MaxHeight,
    MaxWidth,
    MaxLength,
    MaxWeight,
    MaxAxleLoad,
    TimedAllowed,
    TimedDenied,
    DestinationAllowed,
    MaxAxles,
    // Valhalla docs say the max supported value is 31,
    // but the field looks 6 bits wide to me, not 5.
}

#[bitfield(u64)]
#[derive(PartialEq, Eq, FromBytes)]
struct AccessRestrictionBitField {
    #[bits(22)]
    edge_index: u32,
    #[bits(6)]
    restriction_type: u8,
    #[bits(12)]
    modes: u16,
    #[bits(24)]
    _spare: u32,
}

/// Access restrictions beyond the usual access tags
#[derive(PartialEq, Eq, TryFromBytes, Debug)]
#[repr(C)]
pub struct AccessRestriction {
    bitfield: AccessRestrictionBitField,
    value: u64,
}

impl AccessRestriction {
    /// Gets the edge index (within the tile) to which the restriction applies.
    #[inline]
    pub fn edge_index(&self) -> u32 {
        self.bitfield.edge_index()
    }

    /// Gets the type of access restriction.
    ///
    /// # Panics
    ///
    /// This function currently panics if the edge use contains an invalid bit pattern.
    /// We should check the validity of the field when we do the initial try_transmute
    /// on the type (and then have an error result that the tile is invalid).
    /// There are some upstream bugs though between zerocopy and bitfield-struct
    /// macros: https://github.com/google/zerocopy/issues/388.
    #[inline]
    pub fn restriction_type(&self) -> AccessRestrictionType {
        try_transmute!(self.bitfield.restriction_type())
            .map_err(|_| GraphTileError::ValidityError)
            .expect("Invalid bit pattern for restriction type.")
    }

    /// The access modes affected by this restriction.
    #[inline]
    pub fn affected_access_modes(&self) -> EnumSet<Access> {
        // TODO: Look at ways to do this with FromBytes; this currently copies
        // Safety: The access bits are length 12, so invalid representations are impossible.
        unsafe { EnumSet::from_repr_unchecked(self.bitfield.modes()) }
    }
}

#[cfg(test)]
mod tests {
    use super::{Access, AccessRestrictionType};
    use crate::graph_tile::TEST_GRAPH_TILE;
    use enumset::EnumSet;

    #[test]
    fn test_parse_access_restrictions_count() {
        let tile = &*TEST_GRAPH_TILE;

        assert_eq!(
            tile.access_restrictions.len(),
            tile.header.access_restriction_count() as usize
        );
    }

    #[test]
    fn test_parse_access_restrictions() {
        let tile = &*TEST_GRAPH_TILE;

        insta::assert_debug_snapshot!(tile.access_restrictions[0]);

        assert_eq!(
            tile.access_restrictions[0].restriction_type(),
            AccessRestrictionType::MaxWeight
        );
        assert_eq!(
            tile.access_restrictions[0].affected_access_modes(),
            EnumSet::from(Access::Truck)
        );

        insta::assert_debug_snapshot!(tile.access_restrictions.last().unwrap());

        // TODO: Other sanity checks after we add some more advanced methods
    }
}
