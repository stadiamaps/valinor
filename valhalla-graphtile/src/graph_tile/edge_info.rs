use crate::{AsCowStr, BicycleNetwork, graph_tile::GraphTileError, shape_codec::decode_shape};
use bitfield_struct::bitfield;
use enumset::EnumSet;
use geo::LineString;
use std::borrow::Cow;
use std::cell::OnceCell;
use zerocopy::{FromBytes, LE, U16, U32, transmute};
use zerocopy_derive::{FromBytes, Immutable, Unaligned};

#[cfg(feature = "serde")]
use serde::{Serialize, Serializer, ser::SerializeStruct};

#[bitfield(u32,
    repr = U32<LE>,
    from = crate::conv_u32le::from_inner,
    into = crate::conv_u32le::into_inner
)]
#[derive(FromBytes, Immutable, Unaligned)]
pub struct NameInfo {
    #[bits(24, from = crate::conv_u32le::from_inner, into = crate::conv_u32le::into_inner)]
    name_offset: U32<LE>,
    /// Additional fields following the name.
    ///
    /// These can be used for additional information like phonetic readings.
    #[bits(4)]
    additional_fields: u8,
    #[bits(1)]
    is_route_num: u8,
    /// Indicates the string is specially tagged (ex: uses the first char as the tag type).
    ///
    /// This doesn't have any relation to OSM tagging.
    #[bits(1)]
    is_tagged: u8,
    #[bits(2)]
    _spare: u8,
}

#[bitfield(u32,
    repr = U32<LE>,
    from = crate::conv_u32le::from_inner,
    into = crate::conv_u32le::into_inner
)]
#[derive(FromBytes, Immutable, Unaligned)]
struct FirstInnerBitfield {
    #[bits(12, from = crate::conv_u16le::from_inner, into = crate::conv_u16le::into_inner)]
    mean_elevation: U16<LE>,
    #[bits(4)]
    bike_network: u8,
    #[bits(8)]
    speed_limit: u8,
    #[bits(8)]
    extended_way_id: u8,
}

#[bitfield(u32,
    repr = U32<LE>,
    from = crate::conv_u32le::from_inner,
    into = crate::conv_u32le::into_inner
)]
#[derive(FromBytes, Immutable, Unaligned)]
struct SecondInnerBitfield {
    #[bits(4)]
    name_count: u8,
    #[bits(16, from = crate::conv_u16le::from_inner, into = crate::conv_u16le::into_inner)]
    encoded_shape_size: U16<LE>,
    #[bits(8)]
    extended_way_id: u8,
    #[bits(2)]
    extended_way_id_size: u8,
    #[bits(1)]
    has_elevation: u8,
    #[bits(1)]
    _spare: u8,
}

#[derive(Debug, FromBytes, Immutable, Unaligned)]
#[repr(C)]
struct EdgeInfoInner {
    // The first part of the OSM way ID
    // TODO: would be nice to reshuffle this in v4.0
    way_id: U32<LE>,
    first_inner_bitfield: FirstInnerBitfield,
    second_inner_bitfield: SecondInnerBitfield,
}

#[derive(Debug)]
pub struct EdgeInfo<'a> {
    inner: EdgeInfoInner,
    name_info_list: &'a [NameInfo],
    /// The raw varint-encoded shape bytes.
    pub encoded_shape: &'a [u8],
    // Last 2 bytes of the 64-bit way ID
    extended_way_id_2: u8,
    extended_way_id_3: u8,
    decoded_shape: OnceCell<LineString<f64>>,
    // TODO: Final 2 bytes of a 64-bit way ID
    // TODO: Encoded elevation (pointer?)
    text_list_memory: &'a [u8],
    // TODO: Tag cache
}

impl EdgeInfo<'_> {
    /// Gets the tagged speed limit along this edge (in kph).
    #[inline]
    pub const fn speed_limit(&self) -> u8 {
        self.inner.first_inner_bitfield.speed_limit()
    }

    /// Gets the shape of the edge geometry as a [`LineString`].
    ///
    /// # Errors
    ///
    /// See [`decode_shape`] for a description of possible errors.
    ///
    /// # Performance
    ///
    /// If this [`EdgeInfo`] geometry has not been accessed previously,
    /// then it will be decoded during this method call.
    /// Subsequent fetches will be cached for as long as the `EdgeInfo`
    /// is live.
    pub fn shape(&self) -> std::io::Result<&LineString<f64>> {
        // TODO: Use https://doc.rust-lang.org/core/cell/struct.OnceCell.html#method.get_or_try_init when stabilized
        if let Some(linestring) = self.decoded_shape.get() {
            Ok(linestring)
        } else {
            let shape = decode_shape(self.encoded_shape)?;
            Ok(self.decoded_shape.get_or_init(|| shape))
        }
    }

    // TODO: Other filters (tagged and linguistic filters)
    /// Gets all names for this edge.
    ///
    /// # Performance
    ///
    /// This is mostly just pointer indirection and some light filtering.
    /// Not great to call in a hot loop, but also not doing a lot of heavy processing.
    /// The main thing to be careful of in hot paths is allocations.
    pub fn get_names(&self) -> Vec<Cow<'_, str>> {
        self.name_info_list
            .iter()
            .filter_map(|ni| {
                // FIXME: Methods
                // No, this is not a bug...
                if ni.is_tagged() == 0 {
                    Some(self.text_list_memory[ni.name_offset().get() as usize..].as_cow_str())
                } else {
                    None
                }
            })
            .collect()
    }

    /// The bicycle network membership mask for this edge.
    #[inline]
    pub fn bicycle_network(&self) -> EnumSet<BicycleNetwork> {
        // TODO: Look at ways to do this with FromBytes; this currently copies
        // Safety: The access bits are length 4, so invalid representations are impossible.
        unsafe { EnumSet::from_repr_unchecked(self.inner.first_inner_bitfield.bike_network()) }
    }

    /// The way ID of the edge.
    #[inline]
    pub fn way_id(&self) -> u64 {
        u64::from(self.extended_way_id_3) << 56
            | u64::from(self.extended_way_id_2) << 48
            | u64::from(self.inner.second_inner_bitfield.extended_way_id()) << 40
            | u64::from(self.inner.first_inner_bitfield.extended_way_id()) << 32
            | u64::from(self.inner.way_id)
    }
}

// TODO: Feels like this could be a macro
impl<'a> TryFrom<(&'a [u8], &'a [u8])> for EdgeInfo<'a> {
    type Error = GraphTileError;

    fn try_from((bytes, text_list_memory): (&'a [u8], &'a [u8])) -> Result<Self, Self::Error> {
        const INNER_SIZE: usize = size_of::<EdgeInfoInner>();
        let inner_slice: [u8; INNER_SIZE] = (bytes[0..INNER_SIZE]).try_into()?;
        let inner: EdgeInfoInner = transmute!(inner_slice);

        let (name_info_list, bytes) = <[NameInfo]>::ref_from_prefix_with_elems(
            &bytes[INNER_SIZE..],
            inner.second_inner_bitfield.name_count() as usize,
        )
        .map_err(|e| GraphTileError::CastError(e.to_string()))?;

        let offset = 0;
        let (encoded_shape, offset) = {
            let end = offset + inner.second_inner_bitfield.encoded_shape_size().get() as usize;
            (&bytes[offset..end], end)
        };

        // Maybe read a byte; the data structure on disk is tightly packed
        // and drops bytes when possible in exchange for bits that are otherwise unused.
        let (extended_way_id_2, offset) = if inner.second_inner_bitfield.extended_way_id_size() > 0
        {
            (bytes[offset], offset + 1)
        } else {
            (0, offset)
        };

        let (extended_way_id_3, _offset) = if inner.second_inner_bitfield.extended_way_id_size() > 1
        {
            (bytes[offset], offset + 1)
        } else {
            (0, offset)
        };

        // TODO: Elevation

        Ok(Self {
            inner,
            name_info_list,
            encoded_shape,
            extended_way_id_2,
            extended_way_id_3,
            decoded_shape: OnceCell::new(),
            text_list_memory,
        })
    }
}

#[cfg(feature = "serde")]
impl Serialize for EdgeInfo<'_> {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        let num_fields = 4;

        let speed_limit = self.speed_limit();
        let num_fields = if speed_limit == 0 {
            num_fields - 1
        } else {
            num_fields
        };

        let names = self.get_names().join(" / ");
        let num_fields = if names.is_empty() {
            num_fields - 1
        } else {
            num_fields
        };

        let mut state = serializer.serialize_struct("EdgeInfo", num_fields)?;
        if !names.is_empty() {
            state.serialize_field("names", &names)?;
        }

        state.serialize_field(
            "bike_network",
            &self
                .bicycle_network()
                .iter()
                .map(|v| v.as_char())
                .collect::<String>(),
        )?;

        if speed_limit != 0 {
            state.serialize_field("speed_limit", &self.speed_limit())?;
        }

        state.serialize_field("way_id", &self.way_id())?;

        // TODO: elevation

        state.end()
    }
}
