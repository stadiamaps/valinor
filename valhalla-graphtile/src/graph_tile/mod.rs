use thiserror::Error;
use zerocopy::{transmute, try_transmute};

use bytes::Bytes;
use enumset::EnumSet;
#[cfg(test)]
use std::sync::LazyLock;

// To keep files manageable, we will keep internal modules specific to each mega-struct,
// and publicly re-export the public type.
// Each struct should be tested on fixture tiles to a reasonable level of confidence.
// Leverage snapshot tests where possible, compare with Valhalla C++ results,
// and test the first and last elements of variable length vectors
// with snapshot tests.

mod access_restriction;
mod admin;
mod directed_edge;
mod edge_info;
mod header;
mod node;
mod sign;
mod transit;
mod turn_lane;

use crate::{
    graph_id::{GraphId, InvalidGraphIdError},
    transmute_variable_length_data, try_transmute_variable_length_data, Access,
};
pub use access_restriction::{AccessRestriction, AccessRestrictionType};
pub use admin::Admin;
pub use directed_edge::{DirectedEdge, DirectedEdgeExt};
pub use edge_info::EdgeInfo;
pub use header::GraphTileHeader;
pub use node::{NodeInfo, NodeTransition};
pub use sign::{Sign, SignType};
pub use transit::{TransitDeparture, TransitRoute, TransitSchedule, TransitStop, TransitTransfer};
pub use turn_lane::TurnLane;

#[derive(Debug, Error)]
pub enum GraphTileError {
    #[error("Unable to extract a slice of the correct length; the tile data is malformed.")]
    SliceArrayConversion(#[from] std::array::TryFromSliceError),
    #[error("The byte sequence is not valid for this type.")]
    ValidityError,
    #[error("Invalid graph ID.")]
    GraphIdParseError(#[from] InvalidGraphIdError),
}

#[derive(Debug, Error)]
pub enum LookupError {
    #[error("Mismatched base; the graph ID cannot exist in this tile.")]
    MismatchedBase,
    #[error("The feature at the index specified does not exist in this tile.")]
    InvalidIndex,
}

/// A tile within the Valhalla hierarchical tile graph.
pub struct GraphTile {
    memory: Bytes,
    /// Header with various metadata about the tile and internal sizes.
    pub header: GraphTileHeader,
    /// The list of nodes in the graph tile.
    nodes: Vec<NodeInfo>,
    /// The list of transitions between nodes on different levels.
    transitions: Vec<NodeTransition>,
    directed_edges: Vec<DirectedEdge>,
    ext_directed_edges: Vec<DirectedEdgeExt>,
    access_restrictions: Vec<AccessRestriction>,
    transit_departures: Vec<TransitDeparture>,
    transit_stops: Vec<TransitStop>,
    transit_routes: Vec<TransitRoute>,
    transit_schedules: Vec<TransitSchedule>,
    transit_transfers: Vec<TransitTransfer>,
    signs: Vec<Sign>,
    turn_lanes: Vec<TurnLane>,
    admins: Vec<Admin>,
    edge_bins: GraphId,
    // TODO: Complex forward restrictions
    // TODO: Complex reverse restrictions
    // TODO: Street names (names here)
    // TODO: Lane connectivity
    // TODO: Predicted speeds
    // TODO: Stop one stops(?)
    // TODO: Route one stops(?)
    // TODO: Operator one stops(?)
}

impl GraphTile {
    /// Gets the Graph ID of the tile.
    #[inline]
    pub fn graph_id(&self) -> GraphId {
        self.header.graph_id()
    }

    /// Does the supplied graph ID belong in this tile?
    ///
    /// A true result does not necessarily guarantee that an object with this ID exists,
    /// but in that case, either you've cooked up an invalid ID for fun,
    /// or the graph is invalid.
    pub fn may_contain_id(&self, id: &GraphId) -> bool {
        id.tile_base_id() == self.graph_id().tile_base_id()
    }

    /// Gets a reference to a node in this tile with the given graph ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the graph ID cannot be contained in this tile
    /// or the index is invalid.
    pub fn get_node(&self, id: &GraphId) -> Result<&NodeInfo, LookupError> {
        if self.may_contain_id(id) {
            self.nodes
                .get(id.index() as usize)
                .ok_or(LookupError::InvalidIndex)
        } else {
            Err(LookupError::MismatchedBase)
        }
    }

    /// Gets a reference to the directed edge in this tile by graph ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the graph ID cannot be contained in this tile
    /// or the index is invalid.
    pub fn get_directed_edge(&self, id: &GraphId) -> Result<&DirectedEdge, LookupError> {
        if self.may_contain_id(id) {
            self.directed_edges
                .get(id.index() as usize)
                .ok_or(LookupError::InvalidIndex)
        } else {
            Err(LookupError::MismatchedBase)
        }
    }

    /// Gets the opposing edge index, if and only if it exists in this tile.
    pub fn get_opp_edge_index(&self, graph_id: &GraphId) -> Result<u32, LookupError> {
        let edge = self.get_directed_edge(graph_id)?;

        // The edge might leave the tile, so we have to do a complicated lookup
        let end_node_id = edge.end_node_id();
        let opp_edge_index = edge.opposing_edge_index();

        // TODO: Probably a cleaner pattern here?
        let node_edge_index = self.get_node(&end_node_id).map(|n| n.edge_index())?;

        Ok(node_edge_index + opp_edge_index)
    }

    /// Gets a reference to an extended directed edge in this tile by graph ID.
    ///
    /// # Errors
    ///
    /// Returns an error if the graph ID cannot be contained in this tile
    /// or the index is invalid.
    pub fn get_ext_directed_edge(&self, id: &GraphId) -> Result<&DirectedEdgeExt, LookupError> {
        if self.may_contain_id(id) {
            self.ext_directed_edges
                .get(id.index() as usize)
                .ok_or(LookupError::InvalidIndex)
        } else {
            Err(LookupError::MismatchedBase)
        }
    }

    /// Gets access restriction for a directed edge
    /// which apply to *any* of the supplied access modes (ex: auto, bicycle, etc.).
    pub fn get_access_restrictions(
        &self,
        directed_edge_index: u32,
        access_modes: EnumSet<Access>,
    ) -> Vec<&AccessRestriction> {
        // Find the point such that all elements below index are less than directed_edge_index
        // (NOTE: the list is pre-sorted during tile building)
        let index = self
            .access_restrictions
            .partition_point(|e| e.edge_index() < directed_edge_index);

        self.access_restrictions
            .iter()
            // Iterate over the relevant objects (this is an empty iterator if index >= length)
            .skip(index)
            .take_while(|e| e.edge_index() == directed_edge_index)
            .filter(|e| !e.affected_access_modes().is_disjoint(access_modes))
            .collect()
    }

    /// Gets edge info for a directed edge.
    pub fn get_edge_info(&self, directed_edge: &DirectedEdge) -> Result<EdgeInfo, GraphTileError> {
        let edge_info_start = self.header.edge_info_offset as usize;
        let edge_info_offset = directed_edge.edge_info_offset() as usize;

        // TODO: Text slice as a separate property or a function?
        let text_start = self.header.text_list_offset as usize;
        let text_size = self.header.lane_connectivity_offset as usize - text_start;

        Ok(EdgeInfo::try_from((
            self.memory.slice(edge_info_start + edge_info_offset..),
            self.memory.slice(text_start..text_start + text_size),
        ))?)
    }
}

// TODO: Feels like this could be a macro
impl TryFrom<Bytes> for GraphTile {
    type Error = GraphTileError;

    fn try_from(bytes: Bytes) -> Result<Self, Self::Error> {
        let value = bytes.as_ref();
        // Get the byte range of the header so we can transmute it
        const HEADER_SIZE: usize = size_of::<GraphTileHeader>();
        let header_range = 0..HEADER_SIZE;

        // Save the pointer to the list of nodes (the range is consumed)
        let offset = header_range.end;

        // Header
        let header_slice: [u8; HEADER_SIZE] = value[header_range].try_into()?;
        let header: GraphTileHeader = transmute!(header_slice);

        // All the variably sized data arrays

        let (nodes, offset) =
            transmute_variable_length_data!(NodeInfo, value, offset, header.node_count() as usize)?;

        let (transitions, offset) = transmute_variable_length_data!(
            NodeTransition,
            value,
            offset,
            header.transition_count() as usize
        )?;

        let (directed_edges, offset) = try_transmute_variable_length_data!(
            DirectedEdge,
            value,
            offset,
            header.directed_edge_count() as usize
        )?;

        let directed_edge_ext_count = if header.has_ext_directed_edge() {
            header.directed_edge_count() as usize
        } else {
            0
        };
        let (ext_directed_edges, offset) = transmute_variable_length_data!(
            DirectedEdgeExt,
            value,
            offset,
            directed_edge_ext_count
        )?;

        let (access_restrictions, offset) = try_transmute_variable_length_data!(
            AccessRestriction,
            value,
            offset,
            header.access_restriction_count() as usize
        )?;

        let (transit_departures, offset) = transmute_variable_length_data!(
            TransitDeparture,
            value,
            offset,
            header.departure_count() as usize
        )?;

        let (transit_stops, offset) = transmute_variable_length_data!(
            TransitStop,
            value,
            offset,
            header.stop_count() as usize
        )?;

        let (transit_routes, offset) = transmute_variable_length_data!(
            TransitRoute,
            value,
            offset,
            header.route_count() as usize
        )?;

        let (transit_schedules, offset) = transmute_variable_length_data!(
            TransitSchedule,
            value,
            offset,
            header.schedule_count() as usize
        )?;

        let (transit_transfers, offset) = transmute_variable_length_data!(
            TransitTransfer,
            value,
            offset,
            header.transfer_count() as usize
        )?;

        let (signs, offset) =
            try_transmute_variable_length_data!(Sign, value, offset, header.sign_count() as usize)?;

        let (turn_lanes, offset) = try_transmute_variable_length_data!(
            TurnLane,
            value,
            offset,
            header.turn_lane_count() as usize
        )?;

        let (admins, offset) = try_transmute_variable_length_data!(
            Admin,
            value,
            offset,
            header.admin_count() as usize
        )?;

        const U64_SIZE: usize = size_of::<u64>();
        let slice: [u8; U64_SIZE] = (&value[offset..offset + U64_SIZE]).try_into()?;
        let raw_graph_id = transmute!(slice);
        let edge_bins = GraphId::try_from_id(raw_graph_id)?;

        Ok(Self {
            memory: bytes,
            header,
            nodes,
            transitions,
            directed_edges,
            ext_directed_edges,
            access_restrictions,
            transit_departures,
            transit_stops,
            transit_routes,
            transit_schedules,
            transit_transfers,
            signs,
            turn_lanes,
            admins,
            edge_bins,
        })
    }
}

#[cfg(test)]
const TEST_GRAPH_TILE_ID: GraphId = unsafe { GraphId::from_components_unchecked(0, 3015, 0) };

#[cfg(test)]
static TEST_GRAPH_TILE: LazyLock<GraphTile> = LazyLock::new(|| {
    let relative_path = TEST_GRAPH_TILE_ID
        .file_path("gph")
        .expect("Unable to get relative path");
    let path = std::path::PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("fixtures")
        .join("andorra-tiles")
        .join(relative_path);
    let bytes = Bytes::from(std::fs::read(path).expect("Unable to read file"));

    GraphTile::try_from(bytes).expect("Unable to get tile")
});

#[cfg(test)]
mod tests {
    use crate::graph_tile::{TEST_GRAPH_TILE, TEST_GRAPH_TILE_ID};
    use crate::Access;
    use enumset::{enum_set, EnumSet};

    #[test]
    fn test_get_opp_edge_index() {
        let tile = &*TEST_GRAPH_TILE;

        let edges: Vec<_> = (0..u64::from(tile.header.directed_edge_count()))
            .map(|index| {
                let edge_id = TEST_GRAPH_TILE_ID
                    .with_index(index)
                    .expect("Invalid graph ID.");
                let opp_edge_index = tile
                    .get_opp_edge_index(&edge_id)
                    .expect("Unable to get opp edge index.");
                opp_edge_index
            })
            .collect();
        insta::assert_debug_snapshot!(edges);
    }

    #[test]
    fn test_get_access_restrictions() {
        let tile = &*TEST_GRAPH_TILE;

        let mut found_restriction_count = 0;
        let mut found_partial_subset_count = 0;
        for edge_index in 0..tile.header.directed_edge_count() {
            // Sanity check edge index with no access mode filter
            let all_edge_restrictions = tile.get_access_restrictions(edge_index, EnumSet::all());
            for restriction in &all_edge_restrictions {
                found_restriction_count += 1;
                assert_eq!(restriction.edge_index(), edge_index);
            }

            // In the tile fixture, all restrictions apply to trucks
            assert_eq!(
                all_edge_restrictions,
                tile.get_access_restrictions(edge_index, enum_set!(Access::Truck))
            );

            // Empty set should yield no access restrictions
            assert!(tile
                .get_access_restrictions(edge_index, EnumSet::empty())
                .is_empty());

            // This set includes one mode that matches 4 elements in the fixture tile,
            // and another mode that matches none.
            for restriction in
                tile.get_access_restrictions(edge_index, enum_set!(Access::Auto | Access::GolfCart))
            {
                found_partial_subset_count += 1;
                assert_eq!(restriction.edge_index(), edge_index);
            }
        }

        // Hard-coded bits based on the test fixture
        assert_eq!(found_restriction_count, 8);
        assert_eq!(found_partial_subset_count, 4);
    }

    #[test]
    fn test_edge_bins() {
        // TODO: TBH I don't actually understand how this is a valid graph tile. Clearly some black bit magic.
        let tile = &*TEST_GRAPH_TILE;
        assert_eq!(tile.edge_bins.level(), 0);
        assert_eq!(tile.edge_bins.tile_id(), 0);
        assert_eq!(tile.edge_bins.index(), 32000);
    }

    #[test]
    fn test_edge_info() {
        let tile = &*TEST_GRAPH_TILE;
        let first_edge_info = tile
            .get_edge_info(&tile.directed_edges[0])
            .expect("Unable to get edge info.");
        insta::assert_debug_snapshot!("first_edge_info", first_edge_info);
        insta::assert_debug_snapshot!("first_edge_info_decoded_shape", first_edge_info.shape());
        insta::assert_debug_snapshot!("first_edge_info_names", first_edge_info.get_names());

        let last_edge_info = tile
            .get_edge_info(&tile.directed_edges.last().unwrap())
            .expect("Unable to get edge info.");
        insta::assert_debug_snapshot!("last_edge_info", last_edge_info);

        // Edge chosen somewhat randomly; it happens to have multiple names.
        let other_edge_info = tile
            .get_edge_info(&tile.directed_edges[2000])
            .expect("Unable to get edge info.");
        insta::assert_debug_snapshot!("other_edge_info", other_edge_info);
        insta::assert_debug_snapshot!("other_edge_info_names", other_edge_info.get_names());
    }
}
