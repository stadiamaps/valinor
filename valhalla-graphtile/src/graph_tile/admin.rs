use crate::AsCowStr;
use std::borrow::Cow;
use zerocopy::{LE, U32};
use zerocopy_derive::{FromBytes, Immutable, Unaligned};

#[derive(FromBytes, Immutable, Unaligned, Debug)]
#[repr(C)]
pub struct Admin {
    /// The offset into the [`GraphTile`](super::GraphTile) text list for the country name.
    pub country_name_offset: U32<LE>,
    /// The offset into the [`GraphTile`](super::GraphTile) text list for the principal subdivision name.
    pub principal_subdivision_offset: U32<LE>,
    country_iso: [u8; 2],
    principal_subdivision_iso: [u8; 3],
    // padding for alignment
    _spare: [u8; 3],
}

impl Admin {
    /// Gets the ISO 3166-1 country code
    pub fn country_iso(&self) -> Cow<'_, str> {
        self.country_iso.as_cow_str()
    }

    /// Gets the ISO 3166-2 principal subdivision code
    pub fn principal_subdivision_iso(&self) -> Cow<'_, str> {
        self.principal_subdivision_iso.as_cow_str()
    }
}

#[cfg(test)]
mod tests {
    use crate::graph_tile::TEST_GRAPH_TILE;

    #[test]
    fn test_parse_admin_count() {
        let owned_tile = &*TEST_GRAPH_TILE;
        let tile = owned_tile.as_tile();

        assert_eq!(tile.admins.len(), tile.header.admin_count() as usize);
    }

    #[test]
    fn test_parse_admins() {
        let owned_tile = &*TEST_GRAPH_TILE;
        let tile = owned_tile.as_tile();

        let country_isos = tile
            .admins
            .iter()
            .map(|admin| admin.country_iso())
            .collect::<Vec<_>>();
        let principal_subdivision_isos = tile
            .admins
            .iter()
            .map(|admin| admin.principal_subdivision_iso())
            .collect::<Vec<_>>();

        // insta internally does a fork operation, which is not supported under Miri
        if !cfg!(miri) {
            insta::assert_debug_snapshot!("raw_admin_snapshot", tile.admins);
            insta::assert_debug_snapshot!("country_iso", country_isos);
            insta::assert_debug_snapshot!("principal_subdivision_iso", principal_subdivision_isos);
        }
    }
}
