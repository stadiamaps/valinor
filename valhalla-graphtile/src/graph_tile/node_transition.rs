use bitfield_struct::bitfield;
use zerocopy_derive::FromBytes;
use crate::GraphId;

#[derive(FromBytes)]
#[bitfield(u64)]
pub struct NodeTransition {
    #[bits(46)]
    end_node_id: u64,
    // Booleans represented this way for infailability.
    // See comment in node_info.rs for details.
    #[bits(1)]
    up: u8,
    #[bits(17)]
    _spare: u32,
}

impl NodeTransition {
    /// The ID of the corresponding end node on another hierarchy level.
    #[inline]
    pub fn corresponding_end_node_id(&self) -> GraphId {
        // Safety: We know that this value cannot be larger than 46 bits.
        unsafe { GraphId::from_id_unchecked(self.end_node_id()) }
    }

    /// Is the transition up to a higher level?
    /// (If not, there's only one way to go...)
    #[inline]
    pub const fn is_up(&self) -> bool {
        self.up() != 0
    }
}

#[cfg(test)]
mod test {
    use crate::graph_tile::{TEST_GRAPH_TILE};

    #[test]
    fn test_parse_transitions_count() {
        let tile = &*TEST_GRAPH_TILE;

        assert_eq!(tile.transitions.len(), tile.header.transition_count() as usize);
    }

    #[test]
    fn test_parse_transitions() {
        let tile = &*TEST_GRAPH_TILE;

        insta::assert_debug_snapshot!(tile.transitions[0]);
        insta::assert_debug_snapshot!(tile.transitions.last().unwrap());
    }
}