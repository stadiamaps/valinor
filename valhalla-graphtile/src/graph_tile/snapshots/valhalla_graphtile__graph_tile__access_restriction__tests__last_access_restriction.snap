---
source: valhalla-graphtile/src/graph_tile/access_restriction.rs
expression: tile.access_restrictions.last().unwrap()
---
AccessRestriction {
    bitfield: AccessRestrictionBitField {
        edge_index: U32(
            2581,
        ),
        restriction_type: MaxHeight,
        modes: U16(
            233,
        ),
    },
    value: U64(
        430,
    ),
}
